lib/puppet/util/pidlock.rb

```diff
@@ -54,7 +54,21 @@ def clear_if_stale
     begin
       Process.kill(0, lock_pid)
     rescue *errors
-      @lockfile.unlock
+      return @lockfile.unlock
+    end
+
+    # Ensure the process associated with this pid is our process. If
+    # not, we can unlock the lockfile. For now this is only done on
+    # POSIX and Windows platforms (PUP-9247).
+    if Puppet.features.posix?
+      procname = Puppet::Util::Execution.execute(["ps", "-p", lock_pid, "-o", "comm="]).strip
+      args     = Puppet::Util::Execution.execute(["ps", "-p", lock_pid, "-o", "args="]).strip
+      @lockfile.unlock unless procname =~ /ruby/ && args =~ /puppet/ || procname =~ /puppet(-.*)?$/
+    elsif Puppet.features.microsoft_windows?
+      # On Windows, we're checking if the filesystem path name of the running
+      # process is our vendored ruby:
+      exe_path = Puppet::Util::Windows::Process::get_process_image_name_by_pid(lock_pid)
+      @lockfile.unlock unless exe_path =~ /Puppet\\puppet\\bin\\ruby.exe/
```

Comment:
This assumes ruby was installed from AIO. It's possible to download ruby using chocolately or rubyinstaller.org and run `gem install puppet`. Maybe just `\\bin\\ruby.exe$/`?

Reply:
You are right Josh. I tested the code on windows and i installed ruby using choco and puppet as a gem. The path is C:\tools\ruby25\bin so \\bin\\ruby.exe should work.